lib/teleterm/daemon/daemon.go

```diff
@@ -109,6 +109,23 @@ func (s *Service) ResolveCluster(uri string) (*clusters.Cluster, error) {
 	return cluster, nil
 }
 
+// GetCluster returns cluster information
+func (s *Service) GetCluster(ctx context.Context, uri string) (*clusters.Cluster, error) {
+	cluster, err := s.ResolveCluster(uri)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	features, err := cluster.GetClusterFeatures(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cluster.Features = features
```

Comment:
Doesn't it make it harder to understand which fields are available on the struct when looking at a specific line in the codebase?

I can see that there might be problems with exploding combinations in the future and the resource view pattern is something I did see in JSON APIs. But here we have a pretty clear cut situation where in most cases we always want to use a simple version of the object and one endpoint returns all details.

So it seems to me that having two separate structs & messages would make it clearer as to what to expect from specific methods & endpoints that.

Reply:
> Doesn't it make it harder to understand which fields are available on the struct when looking at a specific line in the codebase?

Usually it's not a problem - the caller of RPC X knows what they did and thus which fields are available. It's also common for gRPC clients to deal with missing fields due to version skew (a problem Connect mostly doesn't have).

In this particular case, if the field is important, make it available via more RPCs; otherwise an "optional" approach works.

Good documentation is paramount, though. We need to easily known which RPCs do return this and which don't.

Reply:
I agree that this field wouldn't be needed on `cluster.Cluster` and wouldn't mind removing it. We have `GetClusterFeatures` when needed. 

However, just as `LoggedInUser` is included in the `Cluster` message in the proto package, `Features` should be as well. `LoggedInUser` also only reads from disk stored profile, but this will soon change as we will need to make a network request for `AccessCapabilities`. The only time `LoggedInUser` and `Features` are relevant are for the root cluster as both of these come from the auth server, and can be missing on a leaf cluster query (the only other place this message would return from). 

Merging any cluster information, logged in user information, and auth server info into a single `GetCluster` method (similar to the proposed idea) would be the best place for this. It would also lend itself to a nice cluster init flow in the app when we implement that. 

I'm in favor of removing `Features` from `cluster.Cluster` and  keeping the proposed `features` on the proto Cluster message. We can just update `newApiRootCluster` to include this information. I think it's a good middleground, yes?


edit: `AddCluster` handler also returns a `Cluster` message, but I think this information above is relevant to that call and should be added there too so, doesn't detract.

Reply:
> **In summary:** I'm in favor of _removing_ `Features` from `cluster.Cluster` and _keeping_ the proposed `features` on the proto Cluster message. We can just update `newApiRootCluster` to include this information. I think it's a good middleground. thoughts?

That sounds fine to me.

> edit: `AddCluster` handler also returns a `Cluster` message, but I think this information above is relevant to that call and should be added there too so, doesn't detract.

Yeah, otherwise you'd need to add the root cluster and then sync it immediately, that makes sense. Once we go down that path we might realize that it'd actually be beneficial to add `Features` field on `cluster.Cluster` so do what you think is best.

Reply:
After fiddling with the `AddCluster` side of things, I think it's best to add the `Features` field on `cluster.Cluster` as Alan suggested (apologies for the back and forth). I documented the field, the area in `newApiRootCluster` where it applies, and in proto message. `AddCluster` doesn't (and can't) get any auth server information like features because the user hasn't logged in yet, only added the cluster information to a profile, so the return message will _not_ include the features field (documented). 

I can see `GetCluster` being used quite extensively in the near future with tshd-initiated communication and persistent auth/web clients. Using it to fetch all the relevant information now is a bit wonky, but once those features are in, will be the new paradigm imo.